src/t_hash.c

```diff
@@ -29,52 +29,206 @@
+        sdswrite(buf, field_size, field_sds_type, field, field_len);
+        sdswrite(buf + field_size, buf_size - field_size, SDS_TYPE_8, value, value_len);
+        embedded_field_sds = (sds)(buf + sdsHdrSize(field_sds_type));
+        /* Field sds aux bits are zero, which we use for this entry encoding. */
+        sdsSetAuxBit(embedded_field_sds, FIELD_SDS_AUX_BIT_ENTRY_HAS_PTR_VALUE, 0);
+        serverAssert(entryGetEncoding(embedded_field_sds) == ENTRY_ENC_EMB_VALUE);
+        sdsfree(value);
+    } else {
+        /* Embed field, but not value. Field must be >= SDS_TYPE_8 to encode to
+         * indicate this type of entry.
+         *
+         *     +-------+---------------+
+         *     | value | field         |
+         *     | ptr   | hdr8 "foo" \0 |
+         *     +-------+---------------+
+         */
+        char field_sds_type = sdsReqType(field_len);
+        if (field_sds_type == SDS_TYPE_5) field_sds_type = SDS_TYPE_8;
+        field_size = sdsReqSize(field_len, field_sds_type);
+        size_t alloc_size = sizeof(void *) + field_size;
+        hashTypeEntryPtrValue *entry = zmalloc(alloc_size);
+        entry->value = value;
+        sdswrite(entry->field_data, field_size, field_sds_type, field, field_len);
+        embedded_field_sds = (sds)(entry->field_data + sdsHdrSize(field_sds_type));
+        /* Store the entry encoding type in sds aux bits. */
+        sdsSetAuxBit(embedded_field_sds, FIELD_SDS_AUX_BIT_ENTRY_HAS_PTR_VALUE, 1);
+        serverAssert(entryGetEncoding(embedded_field_sds) == ENTRY_ENC_PTR_VALUE);
+    }
+    return (void *)embedded_field_sds;
 }
 
+/* The entry pointer is the field sds, but that's an implementation detail. */
 sds hashTypeEntryGetField(const hashTypeEntry *entry) {
-    const char *field = entry->field_data + entry->field_offset;
-    return (sds)field;
+    return (sds)entry;
 }
 
 sds hashTypeEntryGetValue(const hashTypeEntry *entry) {
-    return entry->value;
+    switch (entryGetEncoding(entry)) {
+    case ENTRY_ENC_EMB_VALUE: {
+        /* Skip field content, field null terminator and value sds8 hdr. */
+        size_t offset = sdslen(entry) + 1 + sdsHdrSize(SDS_TYPE_8);
+        char *buf = (char *)entry + offset;
+        return buf;
+    }
+    case ENTRY_ENC_PTR_VALUE: {
+        const hashTypeEntryPtrValue *entry_struct = getEntryStruct(entry);
+        return entry_struct->value;
+    }
+    default:
+        serverPanic("Unknown type");
+    }
 }
 
-/* frees previous value, takes ownership of new value */
-static void hashTypeEntryReplaceValue(hashTypeEntry *entry, sds value) {
-    sdsfree(entry->value);
-    entry->value = value;
+/* Returns the address of the entry allocation. */
+static void *hashTypeEntryAllocPtr(hashTypeEntry *entry) {
+    switch (entryGetEncoding(entry)) {
+    case ENTRY_ENC_EMB_VALUE:
+        return sdsAllocPtr(entry);
+    case ENTRY_ENC_PTR_VALUE:
+        return getEntryStruct(entry);
```

Comment:
```suggestion
        return encodedPtrValueEntryGetValueref(entry);
```

Reply:
No, the function name `encodedPtrValueEntryGetValueref` is for finding the value, not the start of the allocation. It's an abuse of the abstraction, which is more confusing than just repeating the calculation here, like

```suggestion
        return sdsAllocPtr(entry) - sizeof(sds *);
```